cmd/plugins/topology-aware/policy/metrics.go

```diff
@@ -0,0 +1,299 @@
+	policyapi "github.com/containers/nri-plugins/pkg/resmgr/policy"
+	"github.com/containers/nri-plugins/pkg/utils/cpuset"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	descZone = iota
+	descZoneCpuSharedCapacity
+	descZoneCpuSharedAssigned
+	descZoneCpuSharedAvailable
+	descZoneMemCapacity
+	descZoneMemAssigned
+	descZoneMemAvailable
+	descZoneContainerCount
+	descZoneSharedContainerCount
+)
+
+var (
+	descriptors = []*prometheus.Desc{
+		descZone: prometheus.NewDesc(
+			"topologyaware_policy_zone_cpu_capacity",
+			"A topology zone of CPUs.",
+			[]string{
+				"zone",
+				"cpus",
+				"mems",
+			},
+			nil,
+		),
+		descZoneCpuSharedCapacity: prometheus.NewDesc(
+			"topologyaware_policy_zone_cpu_shared_capacity",
+			"Capacity of shared CPU pool of a topology zone.",
+			[]string{
+				"zone",
+				"cpus",
+			},
+			nil,
+		),
+		descZoneCpuSharedAssigned: prometheus.NewDesc(
+			"topologyaware_policy_zone_cpu_shared_assigned",
+			"Assigned amount of shared CPU pool of a topology zone.",
+			[]string{
+				"zone",
+				"cpus",
+			},
+			nil,
+		),
+		descZoneCpuSharedAvailable: prometheus.NewDesc(
+			"topologyaware_policy_zone_cpu_shared_available",
+			"Available amount of shared CPU pool of a topology zone.",
+			[]string{
+				"zone",
+				"cpus",
+			},
+			nil,
+		),
+		descZoneMemCapacity: prometheus.NewDesc(
+			"topologyaware_zone_mem_capacity",
+			"Memory capacity of a topology zone.",
+			[]string{
+				"zone",
+				"mems",
+			},
+			nil,
+		),
+		descZoneMemAssigned: prometheus.NewDesc(
+			"topologyaware_zone_mem_assigned",
+			"Amount of assigned memory of a topology zone.",
+			[]string{
+				"zone",
+				"mems",
+			},
+			nil,
+		),
+		descZoneMemAvailable: prometheus.NewDesc(
+			"topologyaware_zone_mem_available",
+			"Amount of available memory of a topology zone.",
+			[]string{
+				"zone",
+				"mems",
+			},
+			nil,
+		),
+		descZoneContainerCount: prometheus.NewDesc(
+			"topologyaware_zone_container_count",
+			"Number of containers assigned to a topology zone.",
+			[]string{
+				"zone",
+			},
+			nil,
+		),
+		descZoneSharedContainerCount: prometheus.NewDesc(
+			"topologyaware_zone_shared_container_count",
+			"Number of containers in the shared CPU pool of a topology zone.",
+			[]string{
+				"zone",
+			},
+			nil,
+		),
+	}
+)
+
+type ZoneMetrics struct {
+	Name             string
+	Cpus             cpuset.CPUSet
+	Mems             libmem.NodeMask
+	SharedPool       cpuset.CPUSet
+	SharedAssigned   int
+	SharedAvailable  int
+	MemCapacity      int64
+	MemAssigned      int64
+	MemAvailable     int64
+	Containers       int
+	SharedContainers int
+}
+
+type TopologyAwareMetrics struct {
+	Zones   []string
+	Metrics map[string]*ZoneMetrics
+}
+
+func (zm *ZoneMetrics) Collect() []prometheus.Metric {
+	if zm == nil {
+		return nil
+	}
+
+	var metrics []prometheus.Metric
+
+	metrics = append(metrics,
+		prometheus.MustNewConstMetric(
```

Comment:
Why the use of "throwaway" const metric? I'd think about initializing a list of persistent gauges and using the Set method to update the value.

Reply:
Because I mindlessly copied the pattern of implementation from balloons without understanding what I am actually doing... Let me try to read up on this a bit more and fix it.